### PR TITLE
Make map&slice with init capacity to decrease grow cost

### PR DESCRIPTION
**Pull request by luckyxiaoqiang**
### Describe what this PR does / why we need it
Make map&slice with init capacity to decrease grow cost.

I have done a simple benchmark of map.  
Benchmark code:
```go
func BenchmarkMap(b *testing.B) {
	for len := 16; len <= 2048; len *= 2 {
		nohint := fmt.Sprintf("map-len-%d-nohint", len)
		b.Run(nohint, func(b *testing.B) {
			keys := stringutil.RandomStrings(10, len)
			for i := 0; i < b.N; i++ {
				m := make(map[string]struct{})
				for _, k := range keys {
					m[k] = struct{}{}
				}
			}
		})

		hint := fmt.Sprintf("map-len-%d-hint", len)
		b.Run(hint, func(b *testing.B) {
			keys := stringutil.RandomStrings(10, len)
			for i := 0; i < b.N; i++ {
				m := make(map[string]struct{}, len)
				for _, k := range keys {
					m[k] = struct{}{}
				}
			}
		})
	}
}
```
Benchmark result:
```
goos: darwin
goarch: amd64
pkg: test
BenchmarkMap/map-len-16-nohint-4                 1024162              1207 ns/op
BenchmarkMap/map-len-16-hint-4                   1979204               612 ns/op
BenchmarkMap/map-len-32-nohint-4                  342871              2920 ns/op
BenchmarkMap/map-len-32-hint-4                    916936              1317 ns/op
BenchmarkMap/map-len-64-nohint-4                  193783              6034 ns/op
BenchmarkMap/map-len-64-hint-4                    428074              2498 ns/op
BenchmarkMap/map-len-128-nohint-4                  99069             11546 ns/op
BenchmarkMap/map-len-128-hint-4                   218790              4932 ns/op
BenchmarkMap/map-len-256-nohint-4                  50898             23931 ns/op
BenchmarkMap/map-len-256-hint-4                   102020             10222 ns/op
BenchmarkMap/map-len-512-nohint-4                  22950             48523 ns/op
BenchmarkMap/map-len-512-hint-4                    57325             19982 ns/op
BenchmarkMap/map-len-1024-nohint-4                 12745             92749 ns/op
BenchmarkMap/map-len-1024-hint-4                   30240             38384 ns/op
BenchmarkMap/map-len-2048-nohint-4                  6195            189864 ns/op
BenchmarkMap/map-len-2048-hint-4                   14448             77945 ns/op
PASS
ok      test    23.131s
```

### Special notes for reviews

### CODE REVIEW

File: core/stat/base/sliding_window_metric.go [hunk 1]
```diff
@@ -159,7 +159,7 @@ func (m *SlidingWindowMetric) SecondMetricsOnCondition(predicate base.TimePredic
 	ws := m.real.ValuesConditional(util.CurrentTimeMillis(), predicate)
 
 	// Aggregate second-level MetricItem (only for stable metrics)
-	wm := make(map[uint64][]*BucketWrap)
+	wm := make(map[uint64][]*BucketWrap, 4)
```
Comment (louyuting) [inline core/stat/base/sliding_window_metric.go:162]: maybe len(ws) is better?

File: core/hotspot/cache/lru.go [hunk 2]
```diff
@@ -31,7 +31,7 @@ func NewLRU(size int, onEvict EvictCallback) (*LRU, error) {
 	c := &LRU{
 		size:      size,
 		evictList: list.New(),
-		items:     make(map[interface{}]*list.Element),
+		items:     make(map[interface{}]*list.Element, size),
```
Comment (louyuting) [inline core/hotspot/cache/lru.go:34]: In default, the size is a so much large number.
Maybe use size doesn't make sense

File: core/stat/base/sliding_window_metric.go (hunk 1 above)
Comment (luckyxiaoqiang) [reply core/stat/base/sliding_window_metric.go:162]: The key changes to `start second`,  many items  in `ws` may have same key in `wm`, it it hard to predict the number of keys in `wm`.

File: core/log/metric/aggregator.go [hunk 3]
```diff
@@ -131,7 +131,7 @@ func isItemTimestampInTime(ts uint64, currentSecStart uint64) bool {
 }
 
 func currentMetricItems(retriever base.MetricItemRetriever, currentTime uint64) map[uint64]*base.MetricItem {
-	m := make(map[uint64]*base.MetricItem, 2)
+	m := make(map[uint64]*base.MetricItem, 8)
```
Comment (sanxun0325) [inline core/log/metric/aggregator.go:134]: maybe len(items) is better?

File: core/hotspot/cache/lru.go (hunk 2 above)
Comment (luckyxiaoqiang) [reply core/hotspot/cache/lru.go:34]: @louyuting Do you mean that it usually don't reach the size?  So what's the recommended hint size?

Comment (luckyxiaoqiang) [reply core/hotspot/cache/lru.go]: I have changed to 64.

File: core/stat/base/leap_array.go [hunk 4]
```diff
@@ -210,7 +210,7 @@ func (la *LeapArray) valuesWithTime(now uint64) []*BucketWrap {
 	if now <= 0 {
 		return make([]*BucketWrap, 0)
 	}
-	ret := make([]*BucketWrap, 0)
+	ret := make([]*BucketWrap, 0, 8)
```
Comment (louyuting) [inline core/stat/base/leap_array.go:213]: maybe `la.array.length` is better?

File: core/stat/base/leap_array.go [hunk 5]
```diff
@@ -225,7 +225,7 @@ func (la *LeapArray) ValuesConditional(now uint64, predicate base.TimePredicate)
...
```
Comment (louyuting) [inline core/stat/base/leap_array.go:228]: maybe la.array.length is better?

File: core/stat/base/leap_array.go (hunk 4 above)
Comment (luckyxiaoqiang) [reply core/stat/base/leap_array.go:213]: Good advice. Length of result mostly is same to  `la.array.length`.

File: core/stat/base/leap_array.go (hunk 5 above)
Comment (luckyxiaoqiang) [reply core/stat/base/leap_array.go:228]: Good advice. Length of result mostly is same to la.array.length.

File: core/log/metric/reader.go [hunk 6]
```diff
@@ -94,7 +94,7 @@ func (r *defaultMetricLogReader) readMetricsInOneFile(filename string, offset ui
 	defer file.Close()
 
 	bufReader := bufio.NewReaderSize(file, 8192)
-	items := make([]*base.MetricItem, 0)
+	items := make([]*base.MetricItem, 0, 8)
```
Comment (louyuting) [inline core/log/metric/reader.go:97]: The metric file  default size is 1024 * 1024 * 50(50M);
From my perspective, we might need to set appropriate size for slice based on file size?

File: core/log/metric/reader.go [hunk 7]
```diff
@@ -129,7 +129,7 @@ func (r *defaultMetricLogReader) readMetricsInOneFileByEndTime(filename string,
...
```
Comment (louyuting) [inline core/log/metric/reader.go:132]: The same with uppon

File: core/log/metric/reader.go (hunk 6 above)
Comment (luckyxiaoqiang) [reply core/log/metric/reader.go:97]: Too many factors do affect, file offset, max lines, preSize.

**Review by louyuting — APPROVED** [review]
Nice work

**Pull request merged**